cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -237,7 +242,48 @@ protected void shutDown() throws Exception {
         result.add(enforceApplicationDetailAccess(entry.getKey(), applicationDetail));
       }
     }
-    return result;
+   return result;
+  }
+
+  public void scanApplications(NamespaceId namespace, Set<String> artifactNames,
+      @Nullable String artifactVersion,
+      Consumer<ApplicationDetail> consumer) {
+
+    Predicate<ApplicationDetail> predicate = getAppPredicate(artifactNames, artifactVersion);
+
+    store.scanApplications(20,
+        (appId, appSpec) -> {
+          if (!appId.getNamespace().equals(namespace.getNamespace())) {
```

Comment:
Why scan across all namespaces? You should add a new method in Store to allow scanning applications in one namespace only.

Reply:
Yes, I am sending out a separate PR for this first.